### DISCUSSION TITLE
Injecting workspace packages is a foot gun

### DISCUSSION THREAD

**Discussion by vjpr** [pnpm/pnpm#3938] [General]
Reference: #3915 

I think this is going to be a foot gun. People are going to use it to solve all their inevitable peer dep issues, but then iterative development workflow is completely broken as you would require `pnpm i` for every change. Any codebase that uses this will be harder to work with.

For every `react` update, you are going to end up supporting multiple versions, hence everyone will just hard-link to avoid fixing things properly.

**Changes**

- Feature should not be recommended unless absolutely necessary, and have a warning in the docs.
- Add a `.npmrc` config option to disable it, or configure it for all packages for testing purposes.

**Alternatives**

- `--experimental-loader` hook in ESM to use the correct version
- Tool to help migrate to a single version of a dep across a codebase (which is what the Googles and Metas do IIRC).
- Watcher to add/remove hardlinks when files added/removed

**Comment by zkochan** [comment]
> Add a .npmrc config option to disable it

Why is it needed? "injected" is set explicitly for every dependency that needs it. We did not add a setting to turn it on for all dependencies.

**Reply by vjpr** [reply]
> "injected" is set explicitly for every dependency that needs it. We did not add a setting to turn it on for all dependencies.

I guess this is a good way to dis-incentivize it, by not allowing global configs.

My use case would be: if I am publishing to npm, before I run tests, i would want to hard-link everything to ensure that peer deps resolve correctly during an install. Maybe this feature already exists, I think I remember something like it... This is why I would like a global option.

I would never see myself using the per-package option. And I guess once someone uses it, there is no way to remove it, because the wrong peers might be used.

**Comment by defjosiah** [comment]
This is semi-related, I had an alternate approach for this same problem:
https://github.com/pnpm/pnpm/issues/3697#issuecomment-908791750 

tl;dr: Use the the explicit "files" property in the package.json to symlink just the files to the proper place (solving the need to re-run an install when files change), and then pull the "dependencies"/"peerDependencies" locally as if it came from the registry (so you get the correct directory structure/peer dependency installation).

If the "files" property changes, you'd need to re-run an install to symlink the new directories/files.

**Reply by zkochan** [reply]
with hard links a re-run is only needed if new files are added to the package.

**Reply by defjosiah** [reply]
Oh awesome! I will check this out
This might be exactly what we want 🎉

**Reply by genshinw** [reply]
> 

I don't know why the hard-linked file not update  especially in vscode. I want to know is there another solution for peer dep issue?

**Comment by guillaumewuip** [comment]
Can we imagine a `pnpm install` flag like `pnpm install --injected` to be used when installing in "injected mode" in a specific package of a monorepo? Having a dedicated flag can be cool to provide a good dev experience for the monorepo. 

In dev mode, we can install with development dependencies. So no need for injected / hard link `workspace:` deps nor `dependenciesMeta.*.injected` in `package.json`. `dependenciesMeta.*.injected` can make local dev a bit complicated because of the need to `pnpm install` again when packages change. 

In production mode (eg. in a Dockerfile) with the `--injected` flag, we can easily install prod deps only and hard link `workspace:` libs.

```
// dev mode
pnpm install 
```

```
// prod mode
pnpm install --injected --prod 
# only prod deps installed
# all "workspace: " dependencies are hard-linked instead of symlink
```

I'm maybe missing something here. Let me know if it's something worth discussing / test or if it's better if i move the discussion in a new Discussion :)

**Comment by grebenyuksv** [comment]
has anyone solved this? I'm thinking of trying to code that watcher which would be re-creating the hard links every time we add files to the hard-linked packages😕

**Reply by genshinw** [reply]
Do you code the watcher? I need it as well 😂

**Reply by grebenyuksv-preply** [reply]
We concluded that the watcher would have to be less trivial than calling `pnpm i` upon file tree modifications. 

The problem is that `pnpm i` seems to be removing and re-creating the hard links to untouched files. Our dev tools weren't able to digest that 🤷‍♀️

**Comment by klingenm** [comment]
I created a utility package that symlinks the files and folders in the root directory of any workspace package that is using `dependenciesMeta.*.injected`. I thought this was a good idea, only requiring a new install after adding or removing a file from the root directory.

In my case, I'm using [rush](https://github.com/microsoft/rushstack/tree/main/apps/rush) to manage the monorepo, keep versions consistent, etc.

This really turned out to be a foot gun. Everything looked as if it worked until running `rush deploy ... ` which simply ignores any symlinks inside node modules package directories.

**Comment by iDVB** [comment]
Anyone have a solve for this?
The issue we are seeing is similar when it comes to our CI
We use Nx to manage our monorepo.

This is our CI:
1. pnpm install
2. build `module1`
3. build `website` (dep on module1 w linkage)

This was failing since `module1`'s dist files get built in step2 after step1's hard linking.
We thought it would be smart to add `pnpm install` to the build command of `module1`
Eg. 
```
scripts: {
  "build": "rollup -c && pnpm install"
}
``` 

But now this issue when there is a **cache hit** for the `module1` build command it, of course, does NOT run the build command and thus also does not run the pnpm install that is tacked on to the end. 😖

**Reply by genshinw** [reply]
I have the same problem, do you find any solution to handle this case?

**Comment by giancarlosisasi** [comment]
has someone found an alternative for the injected feature? 😭

**Comment by NullVoxPopuli** [comment]
I have this plan/idea, but have not had time to implement: https://github.com/pnpm/pnpm/issues/4965#issuecomment-1405264290

**Comment by TimHaboldt** [comment]
Is there any update regarding this issue?

After following way too many threads, searching for solutions all over the internet and stumbling over issues regarding the injected keyword with other [frameworks](https://github.com/pnpm/pnpm/issues/6088) I am revisiting the concern once again.

**Out Setup**
We utilize pnpm as the foundation for our mono repository. It manages all the Node dependencies, and all scripts outside of the Node ecosystem are incorporated into the package.json scripts and can be invoked via pnpm. Our structure differentiates between packages and services. Packages, which contain reusable source code, are used by other packages or services. Services, on the other hand, are deployed into our Kubernetes cluster in the form of deployments, cronjobs, and so forth.

To use the packages in other packages or services, they must first be built since we primarily use TypeScript. Once built, the package can be included via the workspace protocol in the package.json. To enhance the developer experience, we have implemented file watchers for most packages, which can be used in conjunction with hot reloading or file watchers in the services. This accelerates the development process. Changes to a package are synchronized to the node_modules of the service and are detected by the service's file watcher, triggering a rebuild of the service. This efficient setup promotes rapid local development and quick iteration steps.

Moreover, this means that only one version of a package is present at any given time. We currently version very infrequently, and most packages are used directly from the disk, even in production (although the process there is slightly different but not relevant for this discussion).

**Understanding the Issue**

The workspace protocol of pnpm symlinks packages by default. This approach offers an advantage: any modification to a package is immediately synchronized across all other packages or services using it within the mono repository. This enables swift development workflows wherein packages have their own file watcher, and services can automatically hot-reload their website, incorporating the package changes.

However, a significant issue arises: Peer dependencies are completely disregarded, and due to Node's node_modules resolution, this cannot be easily rectified. The problem occurs because Node follows symlinks to their actual location on the disk, which is the folder in your mono repository where your package resides. From there, it reads the source files of your package and starts resolving package imports. These imports are resolved from the context of your mono repository package which in this case is the folder where the package is located. This becomes problematic with peer dependencies.

Peer dependencies are used to inform a package that it needs to have a specific package installed to function, but it doesn't necessarily care about the version of the package. This is crucial for packages like React, where only one React version should be used per webside due to global singletons, etc.

This means that if service A has React 16 installed and another one has React 18 installed, the workspace package using React needs to select a different React version depending on where it's being used. However, due to how Node resolves the symlinks, you only have one package location to work with, and installing both React versions in one node_modules won't work.

To address this, pnpm introduced the dependenciesMeta.inject option, which allows hard-linking packages into the .pnpm folder instead of symlinking them. This generally resolves the issue, as hard-linked packages will not be resolved by Node to their source. This allows multiple copies of the hard-linked package in the .pnpm folder, each using different React versions in their node_modules. Now, packages can have peer dependencies, and the local version is used.

However, this solution presents a new problem: synchronization of file changes.

**The Problem with Hard Links**

Hard links only reference the content of another file. If you're familiar with C++, a hard link is essentially a pointer. It points to the contents of a file. If any pointer to the contents of a file exists, the content will still exist. If all pointers are deleted, the file content will also be deleted. In our case, it means that if you run, for instance, tsc to recompile the dist folder and it recreates the files, the hard links are essentially lost. The files in .pnpm still point to the old file content. Moreover, new files and deleted files are not detected at all. This is why you need to run pnpm install after each package change.

This is very inefficient and slow to work with.

**Moving Forward**

I understand that this issue is not rooted in pnpm and cannot be easily fixed by pnpm. However, I'm curious to know if there are any ideas on how to proceed here, @zkochan. Has anyone created any issues in the Node GitHub repository? Are there any planned updates that might alleviate the pain points with hard links?

This issue is frustrating because it means that pnpm doesn't work with one of the core features of the Node ecosystem. Peer dependencies are used everywhere, and if you want to use packages without publishing them to a registry, you can't avoid them at all. This is evident from the number of issues created regarding peer dependencies. Most of them are completely confused on why this is the matter.

Because other solutions are not very appealing. For example, Nx recommends installing all dependencies of a mono repository in the root package.json of the mono repository [source](https://nx.dev/concepts/more-concepts/dependency-management). We strongly disagree with this approach because it means that every package you want to update requires you to update the entire mono repository. This becomes extremely complex and challenging, especially when packages are using all over the mono repository and changes introduce complete overhauls in their usage.

Solving this issue by not building the package itself and instead building it from the service isn't viable either. First, deletions and file creations are not picked up but can be fixed by doing pnpm install. But we also noticed that file watchers of webpack or nextjs had a lot of troubles picking up the file changes if the content of a hard-linked file got changed. Most of the time, they didn't rebuild or just completed in an unfinished state, which makes our use of hard links practically impossible.

There are even more examples of where this structure is not working. For Turbo, the Nx alternative, it allows you to define [pipelines](https://turbo.build/repo/docs/core-concepts/monorepos/running-tasks). For instance, if you want to build a service, you can define a pipeline that builds all packages required by the service before building the service itself. This process is similar to running a pnpm --filter service-a... build command. It also adds smart caching to the mix.

However, this entire process falls apart because after a package has been built, pnpm install should be executed. Upon conducting some research, it appears that there's no viable solution to this issue in Turbo.

**Conclusion**

Can anyone tell me if there are any solutions out there that make this whole matter easier? Are we using pnpm in combiation with the mono repositroy wrong?
Did anyone else solve this issue while maintaining the fastest development speed possible? 

I invite further comments and suggestions on this topic, as I find it crucial for the pnpm community to find a way to make peer dependencies work.

**Reply by defjosiah** [reply]
I've thought about this issue extensively for work, I think we have the exact same setup as y'all do, especially with this statement.

> Because other solutions are not very appealing. For example, Nx recommends installing all dependencies of a mono repository in the root package.json of the mono repository [source](https://nx.dev/concepts/more-concepts/dependency-management). We strongly disagree with this approach because it means that every package you want to update requires you to update the entire mono repository. This becomes extremely complex and challenging, especially when packages are using all over the mono repository and changes introduce complete overhauls in their usage. 

We built a lot of custom tooling because we couldn't get pnpm, etc. to work with our setup.

My dream for pnpm (and others) after hoping that turbo, nx or pnpm would solve our issues over time would be the following:
1. pnpm provides a "node_modules only" installation, essentially, just let pnpm do what it is best at, installing and putting node_modules in the correct place (using the monorepo), setup each individual packages node_module installation correctly (following peer dependency rules, etc.), but without making any links to the actual built code in the package, essentially, just a bare tree of `node_modules/$package_name/$node_modules` etc.
2. pnpm provides a workspace watcher that monitors for everything in package.json#files, and then copies (no symlink, no hardlink) those files to the corresponding parent directory bare node_modules setup, the individual dependency packages could have their own tsc, etc. running, however they'd like
3. pnpm monitors package.json's for changes since the last "bare" install, and automatically updates the bare checkouts if a dependency is changed, etc.

In my mind, that would provide the best DX for large teams like ours (and yours), who don't want to single version everything in the root, and are constantly running into these issues. We've had to customize nearly all of our webpack, next, etc. configs to prevent peer dependency issues, and we've put off migrating off our custom tooling (which does a worse version of what I just laid out).

Most of us who are doing heavy development on our internal dependencies have file watchers/custom scripts that run and do what I laid out above, but we still constantly run into issues.

Note: 
@NullVoxPopuli implemented something like this in userland here: https://github.com/orgs/pnpm/discussions/3938#discussioncomment-4913170 

But pnpm doing this would be great.

**Reply by NullVoxPopuli** [reply]
direct link to that project: https://github.com/NullVoxPopuli/pnpm-sync-dependencies-meta-injected

**Reply by defjosiah** [reply]
Thanks! I think I meant to paste that link 😂

**Reply by TimHaboldt** [reply]
I've actually seen the package before, popping up in some threads. However, we haven't finished evaluating it internally yet. I'm not sure why I forgot to mention it in the previous post. Thanks for providing the link anyway :smile:.

There are two main reasons why we haven't started using the package yet.

Firstly, it introduces another layer of knowledge that engineers in our company are supposed to have in order to use the mono repository properly. Not only do they need to know that the workspace protocol breaks for peer dependencies, which is already problematic by itself, but they are also required to add another layer on top by implementing the script inside the package.json and referencing it during the start command. It isn't a bad solution, don't get me wrong. I'm really happy that @NullVoxPopuli found a valid solution. But honestly, introducing more complexity to a setup that already seems quite complex is not really desired by us. We might still use it because there aren't any other solutions around though.

The second reason lies in how we want to manage the local services in the future. We are probably going to introduce [skaffold](https://skaffold.dev/docs/) soon, and from what I've experienced personally, Docker and the permissions used inside the container will most likely affect local files. Having the sync package active might affect the .pnpm folder, which will break pnpm locally. I'm not sure if this issue will be the case and it can most likely be prevented (that's why we are still in the evaluation phase of the package :wink:). But it is adding yet another layer of complexity.

I just wish pnpm would introduce something like the package from @NullVoxPopuli, especially because it has been around for a while. If we could natively introduce its functionality into pnpm, it would already eliminate most pain points. @defjosiah's approach of providing a workspace watcher doesn't sound too bad either.

**Reply by TimHaboldt** [reply]
There is also the option of making the developers of node aware of this issue. If they for example provide a way of telling the package resolver from which location they should start resolving packages for symlinked files it would probably already be enough to make this whole Issue go away. 
A check for: if you ever end up in a subfolder of .pnpm. Start resolving packages from this point onward or smth.

**Reply by Ugzuzg** [reply]
~Another issue I have discovered is that the resolution of the dependencies of an injected package doesn't happen or doesn't work correctly. Ideally, it should treat every dependency as if it was installed from a registry.~

Turned out to be an issue with the naive approach that `pnpm-sync-dependencies-meta-injected` takes, not with pnpm. With my dependency tree, the monorepo packages are not only injected into the packages in the repo itself, but also into `node_modules/.pnpm` store, which `pnpm-sync-dependencies-meta-injected` doesn't take into account. If I use `pnpm install` after every single package is built, then the whole thing seems to work.

It's a hard problem to solve in a paradigm of "pnpm is a package manager". It would be easier if the paradigm was "pnpm is a build system".

**Comment by jamesarosen** [comment]
https://github.com/pnpm/pnpm/pull/8857 made `injectWorkspacePackages` required for `pnpm deploy`. I think that's going to push a whole new group of people to set `injectWorkspacePackages: true`.

Because the default (`false`) is such a better development experience, one option would be to set it right before running `deploy`.

Unfortunately, this doesn't work
```sh
pnpm config set inject-workspace-packages=true
pnpm --filter=my-app... install --frozen-lockfile
```
because the `pnpm set` changes the `pnpm-lock.yaml`. That means you have to do

```sh
pnpm config set inject-workspace-packages=true
pnpm --filter=my-app... install --no-frozen-lockfile
```
which disables the version pinning that's so important for secure, confident deploy pipelines.

For now, I recommend avoiding `pnpm@10.x` because it forces this behavior.

**Reply by zkochan** [reply]
You can use `pnpm deploy --legacy` and it will work without the injected workspace packages requirement.

**Reply by netroy** [reply]
I'm seeing `ERR_PNPM_DEPLOY_NONINJECTED_WORKSPACE` on `pnpm deploy` with and without `--legacy` on 10.2.0

**Reply by Aeolun** [reply]
Can confirm that `--legacy` doesn't do anything for me. I'll be the first to admit I don't understand the intricacies of package management, but I'm not really sure why something that's described as `Enables hard-linking of all local workspace dependencies instead of symlinking them.` is relevant to the deploy command?

**Reply by zkochan** [reply]
That should be fixed then. `--legacy` isn't requiring injected dependencies.

> I'll be the first to admit I don't understand the intricacies of package management, but I'm not really sure why something that's described as Enables hard-linking of all local workspace dependencies instead of symlinking them. is relevant to the deploy command?

Users wanted to have the same dependencies installed in the deployed directory as the ones in the workspace. That is a very complicated task without "injected dependencies". In a regular symlinked workspace different projects may have different versions of a peer dependency installed, while during deploy there will be a single version of every peer dependency. So, we can't calculate a deploy lockfile from the workspace lockfile unless we use the injected dependencies feature that kind of groups all projects and isolates them with their sets of peer dependencies. Maybe there are other solutions but this was the easiest one.

**Reply by zkochan** [reply]
The legacy flag works in pnpm 10.2.1

**Reply by Aeolun** [reply]
> So, we can't calculate a deploy lockfile from the workspace lockfile unless we use the injected dependencies feature that kind of groups all projects and isolates them with their sets of peer dependencies. Maybe there are other solutions but this was the easiest one.

Right, so you're saying that sometimes dependencies conflict, and that means you couldn't generate a sensible new lock file from the workspace lock file. If you force a dependency to *always* only be a single version (before or after deploy) then deploy never causes issues, and doesn't even have to care about checking because it'll be enforced beforehand.

That doesn't sound like it has to do with hard or symbolic links though. Is enforcing a single dependency version a side effect of enabling hard links?

This is just curiosity btw, I've resolved my issues.

**Reply by zkochan** [reply]
Yes, if you force a single version of a peer dependency across the workspace, it might be easier to create a lockfile for the deployed project. It has to do with hard links as only with hard links it is possible to properly use multiple versions of a peer dependency in the workspace. You need to hard link the project, so that node.js resolves the peer dependency from a different location. When you symlink, node.js resolves the symlink and always resolves dependencies from a single location.

If we come up with a way to calculate a lockfile without "inject-workspace-packages=true", then we will remove the error message and we will make deploy work without this requirement. Such change will be a minor version bump.

**Reply by kalvenschraut** [reply]
Tried to update to pnpm 10 today and looks like we will be using the --legacy flag. We constantly change files in the workspace packages via a dev server that watches for changes and rebuilds those change. Always hard linking would be really inconvenient with our workflow if we had to reinstall on every change. Currently we use the same workspace version across the whole workspace so we wouldnt have an issue with that.

I am fine adding --legacy to everything to keep the old behavior, but I wanted to make sure it being marked as legacy this option wont be going away without a viable path forward without hard linking permanently. At deploy time I am fine installing as hard link on the CI to build and then reverting back to sym links for not impede on the current dev workflow. alternatively maybe some flag that isnt just "legacy" that people can use if they are fine with the trade offs of the old legacy approach?

Otherwise the update was smooth, thanks for your time and effort into maintaining this!

**Reply by zkochan** [reply]
We will see if we can make it work w/o injected deps. If so, it will stop failing on `inject-workspace-packages=false` and it will be a non-breaking change.

**Reply by fiws** [reply]
I was following the pnpm recipe on monorepos with docker.

I think this setup is broken out of the box because of the change.

https://pnpm.io/docker#example-2-build-multiple-docker-images-in-a-monorepo

**Reply by yuvalkarmi** [reply]
Hey folks, thanks for this. And thanks @zkochan for the awesome package manager.

Adding to this, the default behavior with forcing `inject-workspace-packages=true` makes the DX way worse than it was before.
I need to inject only for the `deploy` command (and I must do so to make my electron project work).
However, passing in the flag just for the deploy fails:
`pnpm deploy --prod --filter desktop-app --config.inject-workspace-packages=true ./desktop-app-dist`

fails with:

`"settings.injectWorkspacePackages" configuration doesn't match the value found in the lockfile`

So now I have to start dealing with hard links in development to ensure that my lockfile doesn't change between dev and the CI pipeline.

Running with `--legacy` doesn't work for me - I do need those hard links - but only for deployment.

Any suggestions?

--------

**EDIT:** pleasantly unexpected, but it seems like if I include a workspace dependency like this in a project's package.json:
`"some-workspace-package": "workspace:*"`

Then it still symlinks! 🎉 And I can still run pnpm deploy.

For anyone using electron, this does the trick with creating a completely flat, accessible node_modules folder with no apparent symlinks for the deploy:

If you don't already have it, add this in your .npmrc at root of project:
`inject-workspace-packages=true`

Run `pnpm install` to update your config, which will install everything in your monorepo

And in the project you want to deploy:
`pnpm deploy --filter name-of-your-workspace-project --shamefully-hoist --config.node-linker=hoisted temp-dist-folder`

**Reply by yuvalkarmi** [reply]
> Yes, if you force a single version of a peer dependency across the workspace, it might be easier to create a lockfile for the deployed project. It has to do with hard links as only with hard links it is possible to properly use multiple versions of a peer dependency in the workspace. You need to hard link the project, so that node.js resolves the peer dependency from a different location. When you symlink, node.js resolves the symlink and always resolves dependencies from a single location.
> 
> If we come up with a way to calculate a lockfile without "inject-workspace-packages=true", then we will remove the error message and we will make deploy work without this requirement. Such change will be a minor version bump.

That would be amazing.

**Reply by segevfiner** [reply]
Offfff. Hit this just now. This sucks... Node.js package management is so much fun...

**Reply by mbrowne** [reply]
I was able to get a Docker build working without using the `--legacy` flag (in my case I already have a build step before the docker build that runs `pnpm install`, so it's not a big deal not to install with a frozen lockfile in the Dockerfile), but I'm not sure how to get this to work together with caching using `pnpm fetch`.

```
pnpm config set inject-workspace-packages=true
pnpm install --prod --filter=my-package --no-frozen-lockfile
pnpm deploy --prod --filter=my-package /prod
```

I can add `pnpm fetch` before that, but not if I also add `--offline` to the `pnpm install` command...and if I understand correctly, the whole point of using `pnpm fetch` is so that you can use `--offline` when installing packages. (The error is, "ERR_PNPM_NO_OFFLINE_TARBALL  A package is missing from the store but cannot download it in offline mode.")

Any ideas on this? Thanks.

**Reply by marczi** [reply]
I have tried different approaches to build a monorepo with Docker's buildx. My measurements show that pnpm fetch may not worth using in monorepos. I use GCP Build and storing/restoring the PNPM store itself via network takes longer than downloading packages. With permanent builder infrastructure it may be different, but my experience is that performance gain is around 10-30 secs with a permanent (mount=type=cache or mount=type=bind) caches from the Docker host. You may try `--prefer-offline` it that solves your problem, however I wanted to use the same `inject-workspace-packages` value both locally and in CI.

**Reply by marczi** [reply]
Also, I highly recommend checking out this blogpost, it helps a lot to understand better monorepo and Docker build tricks and best practices:
https://dev.to/jonlauridsen/exploring-the-monorepo-5-perfect-docker-52aj

**Reply by DASPRiD** [reply]
> **EDIT:** pleasantly unexpected, but it seems like if I include a workspace dependency like this in a project's package.json: `"some-workspace-package": "workspace:*"`
> 
> Then it still symlinks! 🎉 And I can still run pnpm deploy.

While that is true, for some reason you still get hard linking when you remove a package from one of the workspace packages. At that point you have to remove the pnpm-lock.yaml and node_modules folder and do a `pnpm i`.

https://github.com/pnpm/pnpm/issues/9518

**Reply by onlywei** [reply]
> I was able to get a Docker build working without using the `--legacy` flag (in my case I already have a build step before the docker build that runs `pnpm install`, so it's not a big deal not to install with a frozen lockfile in the Dockerfile), but I'm not sure how to get this to work together with caching using `pnpm fetch`.
> 
> ```
> pnpm config set inject-workspace-packages=true
> pnpm install --prod --filter=my-package --no-frozen-lockfile
> pnpm deploy --prod --filter=my-package /prod
> ```
> 
> I can add `pnpm fetch` before that, but not if I also add `--offline` to the `pnpm install` command...and if I understand correctly, the whole point of using `pnpm fetch` is so that you can use `--offline` when installing packages. (The error is, "ERR_PNPM_NO_OFFLINE_TARBALL  A package is missing from the store but cannot download it in offline mode.")
> 
> Any ideas on this? Thanks.

This is amazing insight and I'm going to explore this. I'm curious though: why did you use the flag `--no-frozen-lockfile` in your install command? No such flag is documented: https://pnpm.io/cli/install

**Reply by mbrowne** [reply]
`inject-workspace-packages` causes the lockfile to change. So even though you're not really changing the versions of the packages, it will fail if `frozen-lockfile` is enabled (unless you already had `inject-workspace-packages` in local development, which isn't the norm, nor is it ideal). `frozen-lockfile` is enabled by default for `pnpm install` and `pnpm deploy`. That's why it's important that if you use the code I posted above, you have a step in your CI build process that leaves `frozen-lockfile` _enabled_, otherwise you could run the risk of incorrect versions being used for some of your dependencies—in my case, we have a step to verify the build that runs `pnpm install` _before_ building the Docker image, so I know the lockfile that was committed to git is already good and matches package.json.

As for the missing documentation, you're correct that `--no-frozen-lockfile` isn't documented on the pnpm website, but `--frozen-lockfile` is. I forget where I found info about `--no-frozen-lockfile` but it's just the reverse; I think any boolean flag can be disabled by adding the `no-` prefix.

**Comment by marczi** [comment]
Hi! First, many thanks pnpm, absolutely life-saver. 
I'd like to clarify the expected linking behavior of `inject-workspace-packages=true` in the following setup:

If `packageA` depends on `packageB` via the `workspace:*` protocol (e.g. `"packageB": "workspace:*"` in packageA's package.json), should pnpm create a symlink or a hardlink to packageB in packageA's node_modules?

Using version 10.12.4, I’m seeing a symlink being created (same as @yuvalkarmi reported), but [the documentation](https://pnpm.io/next/settings#injectworkspacepackages) states about injectWorkspacePackages:

    "Enables hard-linking of all local workspace dependencies instead of symlinking them."

@zkochan Is there an official definition of what qualifies as a “local workspace dependency” in this context? Specifically, when using the `workspace:*` protocol in package.json, is pnpm expected to create a symlink instead of a hardlink, even with inject-workspace-packages=true? Or would this be considered a bug?

Thanks! 🙏

**Reply by zkochan** [reply]
There's also another setting, [dedupeInjectedDeps](https://pnpm.io/settings#dedupeinjecteddeps), which is true by default. Hard links for injected deps are in this case only used when they are really needed to make different dependency graphs.

**Comment by guy-signalpet** [comment]
Drawing inspiration from the 
`pnpm config set inject-workspace-packages=true`
I came up with a bigger hack that patches the `pnpm-lock.yaml` and `pnpm-workspace.yaml` files - adding the elusive `injectWorkspacePackages: true` only at build time:

```sh
#!/bin/bash

# Script to inject injectWorkspacePackages: true into pnpm-lock.yaml settings section
# This is a workaround needed for pnpm Docker builds in order to be able to both use a lockfile installation together with the pnpm deploy command -
# while keeping injectWorkspacePackages as false for the local development environment.
# See https://github.com/orgs/pnpm/discussions/3938 and https://pnpm.io/cli/deploy to understand the issue.

set -e

LOCKFILE="pnpm-lock.yaml"

if [ ! -f "$LOCKFILE" ]; then
    echo "Error: $LOCKFILE not found"
    exit 1
fi

# Check if injectWorkspacePackages is already present
if grep -q "injectWorkspacePackages:" "$LOCKFILE"; then
    echo "injectWorkspacePackages already present in $LOCKFILE"
    exit 0
fi

echo "Injecting injectWorkspacePackages: true into $LOCKFILE settings section..."

# Use awk to inject the setting into the settings section
awk '
BEGIN { in_settings = 0; injected = 0 }
/^settings:/ { in_settings = 1; print; next }
in_settings && /^[^ ]/ && !/^settings:/ { 
    if (!injected) {
        print "  injectWorkspacePackages: true"
        injected = 1
    }
    in_settings = 0
    print
    next
}
in_settings && !injected && /^$/ {
    print "  injectWorkspacePackages: true"
    injected = 1
    print
    next
}
{ print }
END {
    if (in_settings && !injected) {
        print "  injectWorkspacePackages: true"
    }
}
' "$LOCKFILE" > "${LOCKFILE}.tmp" && mv "${LOCKFILE}.tmp" "$LOCKFILE"

echo "Successfully injected injectWorkspacePackages: true into $LOCKFILE"
```

Then in the build step:
```dockerfile
FROM base AS build
ARG SERVICE_NAME
WORKDIR /workspace
# Copy the relevant monorepo source code.
COPY . .
# Copy and run script to inject workspace packages setting into lockfile
COPY inject-workspace-packages.sh .
RUN chmod +x inject-workspace-packages.sh && ./inject-workspace-packages.sh && echo "injectWorkspacePackages: true" >> pnpm-workspace.yaml && rm inject-workspace-packages.sh
# Install all dependencies from the cached store for building.
RUN --mount=type=cache,id=pnpm,target=/pnpm/store pnpm install --frozen-lockfile
# Build the service and deploy.
RUN pnpm turbo:build --filter=${SERVICE_NAME}
RUN pnpm deploy --filter=${SERVICE_NAME} --prod /prod/${SERVICE_NAME} 
```

So far it seems to work :)

**Reply by markkorput** [reply]
I was considering a similar solution. @guy-signalpet did you ever run into issues with this workaround?

I'm just wondering why `pnpm install --frozen-lockfile` would fail on the setting mismatch in the first place if it could be simply ignored?

**Reply by guy-signalpet** [reply]
Nope its working perfectly for us. The reason it would fail is (I guess) because pnpm compares the lockfiles as-is without exclusions, which makes a lot of sense from a complexity perspective. The real issue for us is not being able to use pnpm deploy in Dockerfiles when injectWorkspacePackages is off.

Turning injectWorkspacePackages on by default is a no-go since it completely defeats the purpose (imo) of using pnpm with a tool like Turborepo in order to enable fast incremental builds. Perhaps a good tradeoff is for pnpm deploy to be able to "auto switch" this setting on just for the purpose of the command? Im not really sure how it works internally so I may be talking nonsense :)

**Reply by obeattie** [reply]
Just a tip: this is more portable, and will work in eg. Alpine containers, if you change the shebang to `#!/bin/sh`. (The script is POSIX-compatible so it works fine under `sh`)
